File: src/main/java/toby/command/commands/moderation/KickCommand.java
package toby.command.commands.moderation;

import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Mentions;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.api.interactions.commands.build.OptionData;
import toby.command.CommandContext;
import toby.jpa.dto.UserDto;

import java.util.List;
import java.util.Optional;

import static toby.command.ICommand.deleteAfter;
import static toby.command.ICommand.invokeDeleteOnMessageResponse;

public class KickCommand implements IModerationCommand {

    private final String USERS = "users";

    @Override
    public void handle(CommandContext ctx, UserDto requestingUserDto, Integer deleteDelay) {
        deleteAfter(ctx.getEvent().getHook(), deleteDelay);
        final SlashCommandInteractionEvent event = ctx.getEvent();
        event.deferReply().queue();
        final Guild guild = event.getGuild();
        final Member member = ctx.getMember();
        final Member botMember = guild.getSelfMember();

        Optional<List<Member>> optionalMemberList = Optional.ofNullable(event.getOption(USERS)).map(OptionMapping::getMentions).map(Mentions::getMembers);
        if (optionalMemberList.isEmpty()) {
            event.getHook().sendMessage("You must mention 1 or more Users to shoot").queue(invokeDeleteOnMessageResponse(deleteDelay));
            return;
        }

        optionalMemberList.get().forEach(target -> {

            if (!member.canInteract(target) || !member.hasPermission(Permission.KICK_MEMBERS)) {
                event.getHook().sendMessageFormat("You can't kick %s", target).queue(invokeDeleteOnMessageResponse(deleteDelay));
                return;
            }


            if (!botMember.canInteract(target) || !botMember.hasPermission(Permission.KICK_MEMBERS)) {
                event.getHook().sendMessageFormat("I'm not allowed to kick %s", target).queue(invokeDeleteOnMessageResponse(deleteDelay));
                return;
            }

            guild.kick(target).reason("because you told me to.").queue(
                    (__) -> event.getHook().sendMessage("Shot hit the mark... something about fortnite?").queue(invokeDeleteOnMessageResponse(deleteDelay)),
                    (error) -> event.getHook().sendMessageFormat("Could not kick %s", error.getMessage()).queue(invokeDeleteOnMessageResponse(deleteDelay)));
        });
    }

    @Override
    public String getName() {
        return "kick";
    }

    @Override
    public String getDescription() {
        return "Kick a member off the server.";

    }

    @Override
    public List<OptionData> getOptionData() {
        return List.of(new OptionData(OptionType.STRING, USERS, "User(s) to kick", true));
    }
}

File: src/main/kotlin/toby/command/commands/moderation/KickCommand.kt
package toby.command.commands.moderation

import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.Mentions
import net.dv8tion.jda.api.interactions.commands.OptionMapping
import net.dv8tion.jda.api.interactions.commands.OptionType
import net.dv8tion.jda.api.interactions.commands.build.OptionData
import toby.command.CommandContext
import toby.command.ICommand.Companion.invokeDeleteOnMessageResponse
import toby.jpa.dto.UserDto

class KickCommand : IModerationCommand {
    companion object {
        private const val USERS = "users"
    }

    override fun handle(ctx: CommandContext, requestingUserDto: UserDto, deleteDelay: Int?) {
        val event = ctx.event
        val deleteDelaySafe = deleteDelay ?: 0
        event.deferReply().queue()

        val guild = event.guild!!
        val member = ctx.member ?: return
        val botMember = guild.selfMember

        val mentionedMembers = event.getOption(USERS)
            ?.mentions
            ?.members

        if (mentionedMembers.isNullOrEmpty()) {
            event.hook.sendMessage("You must mention 1 or more Users to kick.").queue(invokeDeleteOnMessageResponse(deleteDelaySafe))
            return
        }

        mentionedMembers.forEach { target ->
            when {
                !member.canInteract(target) || !member.hasPermission(Permission.KICK_MEMBERS) -> {
                    event.hook
                        .sendMessage("You can't kick ${target.effectiveName}")
                        .queue(invokeDeleteOnMessageResponse(deleteDelaySafe))
                }
                !botMember.canInteract(target) || !botMember.hasPermission(Permission.KICK_MEMBERS) -> {
                    event.hook
                        .sendMessage("I'm not allowed to kick ${target.effectiveName}")
                        .queue(invokeDeleteOnMessageResponse(deleteDelaySafe))
                }
                else -> {
                    guild.kick(target).reason("because you told me to.").queue(
                        { event.hook.sendMessage("Shot hit the mark... something about fortnite?").queue(invokeDeleteOnMessageResponse(deleteDelaySafe)) },
                        { error -> event.hook.sendMessage("Could not kick: ${error.message}").queue(invokeDeleteOnMessageResponse(deleteDelaySafe)) }
                    )
                }
            }
        }
    }

    override val name: String
        get() = "kick"

    override val description: String
        get() = "Kick a member off the server."

    override val optionData: List<OptionData>
        get() = listOf(
            OptionData(OptionType.STRING, USERS, "User(s) to kick", true)
        )
}
